contract-examples/test/utils.ts

```diff
@@ -0,0 +1,123 @@
+test.debug = (name, fnNameOrObject, overrides = {}) => it.only(name, testFn(fnNameOrObject, overrides, true))
+// `test.skip` is used very similarly to `it.skip` from Mocha, it will skip all tests marked with `test.skip`
+test.skip = (name, fnNameOrObject, overrides = {}) => it.skip(name, testFn(fnNameOrObject, overrides))
+
+// `testFn` is a higher-order function. It returns a function that can be used as the test function for `it`
+const testFn = (fnNameOrObject: FnNameOrObject, overrides = {}, debug = false) => {
+  // normalize the input to an array of objects
+  const fnObjects: MethodWithDebugAndOverrides[] = (Array.isArray(fnNameOrObject) ? fnNameOrObject : [fnNameOrObject]).map(fnNameOrObject => {
+    fnNameOrObject = typeof fnNameOrObject === 'string' ? { method: fnNameOrObject } : fnNameOrObject
+    // assign all default values and overrides
+    fnNameOrObject.overrides = Object.assign({}, overrides, fnNameOrObject.overrides ?? {})
+    fnNameOrObject.debug = fnNameOrObject.debug ?? debug
+    fnNameOrObject.shouldFail = fnNameOrObject.shouldFail ?? false
+
+    return fnNameOrObject as MethodWithDebugAndOverrides
+  })
+
+  // only `test_` prefixed functions can be called on the `DSTest` contracts to clearly separate tests and helpers
+  assert(fnObjects.every(({ method }) => method.startsWith('test_')), "Solidity test functions must be prefixed with 'test_'")
+
+  // return the test function that will be used by `it`
+  // this function must be defined with the `function` keyword so that `this` is bound to the Mocha context
+  return async function() {
+    // `Array.prototype.reduce` is used here to ensure that the test functions are called in order.
+    // Each test function waits for its predecessor to complete before starting
+    return fnObjects.reduce((p: Promise<undefined>, fn) => p.then(async () => {
+      const contract = fn.overrides.from
+        ? this.testContract.connect(await ethers.getSigner(fn.overrides.from))
+        : this.testContract
+      const tx = await contract[fn.method](fn.overrides).catch(err => {
+        if (fn.shouldFail) {
+          if (fn.debug) console.error(`smart contract call failed with error:\n${err}\n`)
+
+          return { failed: true }
+        }
+
+        console.error("smart contract call failed with error:", err)
+        throw err
+      })
+
+      // no more assertions necessary if the method-call should fail and did fail
+      if (tx.failed && fn.shouldFail) return
+
+      const txReceipt = await tx.wait().catch(err => {
+        if (fn.debug) console.error(`tx failed with error:\n${err}\n`)
+        return err.receipt
+      })
+
+      // `txReceipt.status` will be `0` if the transaction failed.
+      // `contract.callStatic.failed()` will return `true` if any of the `DSTest` assertions failed.
+      const failed = txReceipt.status === 0 ? true : await contract.callStatic.failed()
+      if (fn.debug || failed) {
+        console.log('')
```

Comment:
Is this console log necessary?

Reply:
And where am I supposed to put the `\n` character?